File: object-oriented-programming/ObjectOrientedProgramming.java
// Interface
interface IAnimal {
    String speak();
    String describe();
}

// Abstract class
abstract class Animal implements IAnimal {
    protected String name;

    public Animal(String name) {
        this.name = name;
    }

    @Override
    abstract public String speak();

    @Override
    public String describe() {
        return "I am " + name;
    }
}

// Concrete class: Dog
class Dog extends Animal {
    public Dog(String name) {
        super(name);
    }

    @Override
    public String speak() {
        return name + " says: Woof!";
    }
}

// Concrete class: Cat
class Cat extends Animal {
    public Cat(String name) {
        super(name);
    }

    @Override
    public String speak() {
        return name + " says: Meow!";
    }
}

public class ObjectOrientedProgramming {
    public static void main(String[] args) {
        System.out.println("Language: Java");

        IAnimal[] animals = {new Dog("Rex"), new Cat("Whiskers")};

        for (IAnimal animal : animals) {
            System.out.println(animal.speak());
            System.out.println(animal.describe());
        }
    }
}

File: object-oriented-programming/ObjectOrientedProgramming.cs
using System;
using System.Collections.Generic;

// Interface
interface IAnimal
{
    string Speak();
    string Describe();
}

// Abstract class
abstract class Animal : IAnimal
{
    protected string name;

    public Animal(string name)
    {
        this.name = name;
    }

    public abstract string Speak();

    public virtual string Describe()
    {
        return $"I am {name}";
    }
}

// Concrete class: Dog
class Dog : Animal
{
    public Dog(string name) : base(name) { }

    public override string Speak()
    {
        return $"{name} says: Woof!";
    }
}

// Concrete class: Cat
class Cat : Animal
{
    public Cat(string name) : base(name) { }

    public override string Speak()
    {
        return $"{name} says: Meow!";
    }
}

class ObjectOrientedProgramming
{
    static void Main()
    {
        Console.WriteLine("Language: C#");

        List<IAnimal> animals = new List<IAnimal>
        {
            new Dog("Rex"),
            new Cat("Whiskers")
        };

        foreach (var animal in animals)
        {
            Console.WriteLine(animal.Speak());
            Console.WriteLine(animal.Describe());
        }
    }
}
